### PR TITLE
WebTransport API

### PR DESCRIPTION
**Pull request by hamishwillee**
The Web Transport API has been through some spec updates (currently supported only in FF nightly). This documents the changes.

- [x] `WebTransport` constructor - new options: `allowPooling`, `requireUnreliable`, `congestionControl`
- [x] `WebTransport.reliability` - new property
- [x] `WebTransport.draining` - new property
- [x] `WebTransport.getStats()` - new method
- [x] `WebTransport.congestionControl` - new property
- [x] `WebTransport.congestionControl.createBidirectionalStream()` - new option argument.
- [x]  `WebTransport.congestionControl.createUnidirectionalStream ()` - new option argument.
- [x]  `WebTransportReceiveStream` - missing
  - [x] `WebTransportReceiveStream.getStats()` method 

Other docs work for this can be tracked in #26153

### CODE REVIEW

File: files/en-us/web/api/webtransport/webtransport/index.md [hunk 1]
```diff
@@ -10,39 +10,62 @@ browser-compat: api.WebTransport.WebTransport
...
+      - : A string indicating the application's preference that the congestion control algorithm used when sending data over this connection be tuned for either throughput or low-latency.
+        This is a hint to the user agent.
+        The allowed values are: `default` (default), `throughput`, and `low-latency`.
+    - `requireUnreliable` {{optional_inline}}
+      - : A boolean value.
+        If `true`, the connection cannot be established over HTTP/2 if an HTTP/3 connection is not possible.
+        By default the value is `false`.
     - `serverCertificateHashes` {{optional_inline}}
-      - : An array of `WebTransportHash` objects. If specified, it allows the website to connect to a server by authenticating the certificate against the expected certificate hash instead of using the Web public key infrastructure (PKI). This feature allows Web developers to connect to WebTransport servers that would normally find obtaining a publicly trusted certificate challenging, such as hosts that are not publicly routable, or ephemeral hosts like virtual machines.
+      - : An array of objects defining a hash value and its associated algorithm.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransport/webtransport/index.md:45]: I find `serverCertificateHashes` confusing - can anyone confirm the following?

Firstly, what normally happens with PKI is that the browser connects to a server. The server sends its certificate which is signed by a CA. The user agent has a bunch of public certificates that it can use to verify that the certificate was signed by a particular CA, and hence its information can be trusted.

With `serverCertificateHashes` the code connecting to the server sends algorithm names and hashes, but it isn't clear how this works - how is the trust generated?

My "guess" is that the site has a certificate, and the user code served by that site can at some point generate a hash of the certificate using one or more algorithms.  So when the user agent connects to the site it gets a certificate, and when it gets the constructor call it can check that the hash matches - i.e. the user code trying to use the transport and the server certificate must come from the same origin?

Doesn't that mean that the server would have to be written to serve user code that is updated with the current hash information? 

Why is this useful? - to set up the HTTPS connection didn't you have to set up PKI in the first place? The stated use of the mechanism to to allow developers to connect to servers that find it hard to get a publically trusted certificate (i.e. isn't this a catch 22)

Am I being dim? This doesn't seem obvious from the text.

File: files/en-us/web/api/webtransport/webtransport/index.md [hunk 2]
```diff
@@ -10,39 +10,62 @@ browser-compat: api.WebTransport.WebTransport
...
+        - `value`
+          - : A `BufferSource` representing the hash value.
 
 ### Exceptions
 
 - `NotSupportedError` {{domxref("DOMException")}}
   - : Thrown if `serverCertificateHashes` is specified but the transport protocol does not support this feature.
 - `SyntaxError` {{domxref("DOMException")}}
   - : Thrown if the specified `url` is invalid, if the scheme is not HTTPS, or if the URL includes a fragment.
+- `TypeError` {{domxref("DOMException")}}
```
Comment (teoli2003) [inline files/en-us/web/api/webtransport/webtransport/index.md:67]: Drive-by comment. I think both `SyntaxError` and `TypeError` maybe JavaScript exceptions and not `DOMException` objects. Can you double-check?

Comment (hamishwillee) [reply files/en-us/web/api/webtransport/webtransport/index.md:67]: Yes, I think you're right. I just copied the pattern.

File: files/en-us/web/api/webtransport/reliability/index.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+title: "WebTransport: reliability property"
+short-title: reliability
+slug: Web/API/WebTransport/reliability
+page-type: web-api-instance-property
+status:
+  - experimental
+browser-compat: api.WebTransport.reliability
+---
+
+{{APIRef("WebTransport API")}}{{SeeCompatTable}}{{SecureContext_Header}}
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransport/reliability/index.md:11]: All of the properties/methods have available in workers and secure context headers. I have just followed the pattern. Don't we just add secure context to the entry points? But keep the AvailableInWorkers on every relevant page?

Note, I actually prefer securecontext to be added on every page, but I'd rather it was done as metadata and rendered automatically as a much smaller icon.




I like the pattern but I think t

File: files/en-us/web/api/webtransport/draining/index.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+slug: Web/API/WebTransport/draining
+page-type: web-api-instance-property
+status:
+  - experimental
+browser-compat: api.WebTransport.draining
+---
+
+{{APIRef("WebTransport API")}}{{SeeCompatTable}}{{SecureContext_Header}}
+
+The **`draining`** read-only property of the {{domxref("WebTransport")}} interface returns a promise that resolves if a server indicates that it wants the client to gracefully close the connection.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransport/draining/index.md:13]: FYI The spec is pretty unhelpful about what "draining" really means, forwarding you to HTTP spec, which is equally opaque. Fortunately I found this https://github.com/w3c/webtransport/issues/432

Essentially draining combines two signals in the HTTP layer - one to say "go away" meaning, I want not more sessions but you can continue your current ones, and another to say "start gracefully shutting yourself down".

The term "meaning" itself isn't provided anywhere AFAIK but I think it implies "drain out the last bit of info you need to do so safely and then shut down". Either end can send this in HTTP, but in WebTransport it only gets sent from the server end.

File: files/en-us/web/api/webtransport/reliability/index.md (hunk 3 above)
Comment (Elchi3) [reply files/en-us/web/api/webtransport/reliability/index.md:11]: I usually only add it to the interface pages and entry point pages. That's aligned with how it is written in IDL, too, I think.
Following the existing pattern seems OK, too, though.

File: files/en-us/web/api/webtransportdatagramduplexstream/outgoingmaxage/index.md [hunk 5]
```diff
@@ -1,6 +1,7 @@
 ---
 title: WebTransportDatagramDuplexStream.outgoingMaxAge
```
Comment (Elchi3) [inline files/en-us/web/api/webtransportdatagramduplexstream/outgoingmaxage/index.md:2]:
```suggestion
title: WebTransportDatagramDuplexStream: outgoingMaxAge property
short-title: outgoingMaxAge
```

File: files/en-us/web/api/webtransport/reliability/index.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+The **`reliability`** read-only property of the {{domxref("WebTransport")}} interface indicates whether the connection supports unreliable transports, such as UDP, or will only run over reliable transports.
```
Comment (Elchi3) [inline files/en-us/web/api/webtransport/reliability/index.md:13]: This doesn't read so clear to me. 

```suggestion
The **`reliability`** read-only property of the {{domxref("WebTransport")}} interface indicates whether the connection supports unreliable transports, such as UDP, or whether it will run over reliable transports only.
```

File: files/en-us/web/api/webtransport/index.md [hunk 7]
```diff
@@ -32,6 +32,10 @@ The **`WebTransport`** interface of the {{domxref("WebTransport API", "WebTransp
   - : Represents one or more unidirectional streams opened by the server. Returns a {{domxref("ReadableStream")}} of {{domxref("WebTransportReceiveStream")}} objects. Each one can be used to read data from the server.
 - {{domxref("WebTransport.ready", "ready")}} {{ReadOnlyInline}} {{Experimental_Inline}}
   - : Returns a promise that resolves when the transport is ready to use.
+- {{domxref("WebTransport.reliability", "reliability")}} {{ReadOnlyInline}} {{Experimental_Inline}}
+  - : Returns a string that indicates whether the connection supports unreliable transports, such as UDP, or will only run over reliable transport.
```
Comment (Elchi3) [inline files/en-us/web/api/webtransport/index.md:36]: See my other suggestion for maybe rewording this slightly.

File: files/en-us/web/api/webtransport/webtransport/index.md (hunk 1 above)
Comment (Elchi3) [reply files/en-us/web/api/webtransport/webtransport/index.md:45]: I don't know this well enough. Might be worth asking a SME.

> Why is this useful?

Maybe for this?(https://w3c.github.io/webtransport/#certificate-hashes):

"The WebTransport API lets the application specify multiple certificate hashes at once, allowing the client to accept multiple certificates for a period in which a new certificate is being rolled out."

> how is the trust generated?

https://w3c.github.io/webtransport/#certificate-hashes might give some answers too

File: files/en-us/web/api/webtransport/reliability/index.md (hunk 3 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransport/reliability/index.md:11]: Thanks. I'm going to follow the site pattern and remove these.

File: files/en-us/web/api/webtransport/reliability/index.md (hunk 6 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransport/reliability/index.md:13]: Thank you. I like the second option.

File: files/en-us/web/api/webtransport/index.md (hunk 7 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransport/index.md:36]:
```suggestion
  - : Returns a string that indicates whether the connection supports reliable transports only, or whether it also supports unreliable transports (such as UDP).

File: files/en-us/web/api/webtransport/webtransport/index.md (hunk 1 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransport/webtransport/index.md:45]: Thanks. That's somewhat helpful. I have asked on the spec: https://github.com/w3c/webtransport/issues/508

There will be other things to note here. Currently only SHA-256 is supported in spec as an algorithm.

File: files/en-us/web/api/webtransport/webtransport/index.md [hunk 8]
```diff
@@ -8,59 +8,95 @@ status:
...
+          - : A string with the value: `sha-256` (case-insensitive).
+            Note that this string represents the algorithm to use to verify the hash, and that any hash using an unknown algorithm will be ignored.
+            Currently the only "known" algorithm in the spec is `sha-256`.
 
-- `algorithm`
-  - : A string representing the algorithm to use to verify the hash. Any hash using an unknown algorithm will be ignored.
-- `value`
-  - : A `BufferSource` representing the hash value.
+        - `value`
+          - : An [`ArrayBuffer`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) or {{domxref("ArrayBufferView")}} containing the hash value.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransport/webtransport/index.md:63]: I'm a bit confused about this too. Specifically, this takes a `BufferSource` which is an `ArrayBuffer` or an `ArrayBufferView`. However I have been told that this will work if those hashes match a cert. That implies that a string can implicitly map direct to an ArrayBuffer - but I don't see that anywhere?

```
const transport = new WebTransport(url, {
  serverCertificateHashes: [
    {
      algorithm: "sha-256",
      value: "5a155927eba7996228455e4721e6fe5f739ae15db6915d765e5db302b4f8a274",
    }
  ],
});
```

File: files/en-us/web/api/webtransportreceivestream/getstats/index.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  - experimental
+browser-compat: api.WebTransportReceiveStream.getStats
+---
+
+{{APIRef("WebTransport API")}}{{SeeCompatTable}}
+
+The **`getStats()`** method of the {{domxref("WebTransportReceiveStream")}} interface asynchronously returns an object containing statistics for the current stream.
+
+The statistics include the total number of ordered bytes that have arrived on this stream (ignoring network overhead, up until the first missing byte) and the total number that have been read by the application.
+It therefore provides a measure of how quickly the application is consuming bytes from the server on this particular stream.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransportreceivestream/getstats/index.md:16]: FYI This is an educated guess on the purpose of the API. You get the incoming bytes in stream and the amount read, which tells you what is on the queue, and by timestamps you could estimate the rate of consumption. The bit "up until the first missing byte" is I think because the bytes can arrive out of order - so the stream won't count any subsequent bytes it might receive because they can't yet be consumed.

File: files/en-us/web/api/webtransportreceivestream/getstats/index.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+- `bytesReceived`
+  - : A positive integer indicating the number of bytes received by this stream, up to the first missing byte.
+     The number does not include any network overhead, and can only increase.
+- `bytesRead`
+  - : A positive integer indicating the number of bytes the application has read from this `WebTransportReceiveStream` stream.
+    This number can only increase, and is always less than or equal to `bytesReceived`.
+
+## Examples
+
+The code snippet below uses `await` to wait on the {{jsxref("Promise")}} returned by `getStats()`.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransportreceivestream/getstats/index.md:46]: I haven't got any real way to test this. So this example is simplistic. It won't be incorrect, but of course not as good as something more complete. A live example would be nice.

File: files/en-us/web/api/webtransportreceivestream/index.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+slug: Web/API/WebTransportReceiveStream
+page-type: web-api-interface
+status:
+  - experimental
+browser-compat: api.WebTransportReceiveStream
+---
+
+{{APIRef("WebTransport API")}}{{SeeCompatTable}}
+
+The `WebTransportReceiveStream` interface of the {{domxref("WebTransport API", "WebTransport API", "", "nocode")}} is a {{domxref("ReadableStream")}} that can be used to read from an incoming unidirectional or bidirectional {{domxref("WebTransport")}} stream from a server.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransportreceivestream/index.md:12]: The spec says this can read both unidirectional and bidirectional incoming streams. However there is a different property and associated interface for bidirectional streams that you can read/write from, so I'm not clear on whether this is "true" or some kind of typo. Asked here https://bugzilla.mozilla.org/show_bug.cgi?id=1818754#c14 but I wouldn't block on this.

File: files/en-us/web/api/webtransport/webtransport/index.md (hunk 1 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransport/webtransport/index.md:47]: OK, I have added the certificate restrictions. I now think the way this works is that the user agent mediates the TLS discussion with the server. So the author of code using the hashes may have nothing but the hashes which they supply to the API. The server gets the certificate from the site during the TLS handshake so it can hash it and compare to the one supplied in the constructor. That tells it the server has this certificate, which is all that TLS itself can do.

What isn't clear is if I'm right, and if I am, how the developer is expected to get this hash in the first place. I mean the spec authors must have an idea. But I can't get a straight answer to that bit. I'm still hoping for info from FF team to confirm too.

Upshot though, this should not block us, because it is no worse than currently (and significantly better in many ways).

File: files/en-us/web/api/webtransport/webtransport/index.md [hunk 12]
```diff
@@ -72,6 +118,45 @@ async function closeTransport(transport) {
...
+```js
+const transport = new WebTransport(url, {
+  serverCertificateHashes: [
+    {
+      algorithm: "sha-256",
+      value: "5a155927eba7996228455e4721e6fe5f739ae15db6915d765e5db302b4f8a274",
+    },
+    {
+      algorithm: "sha-256",
+      value: "7d7094e7a8d3097feff3b5ee84fa5cab58e4de78f38bcfdee5ea8b51f4bfa8fd",
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransport/webtransport/index.md:154]: Spec author initially said this was good, but then on follow up said "WebRTC accepts strings, but it looks like this one needs conversion.". I'm hoping our engineers can confirm. The thing is that this value takes an ArrayBuffer or one of the typed arrays. So it might be that this needs to show TextEncoder string to buffer conversion or similar. Hoping for direction. Not sure if this would count as a "Blocker".

**Comment by hamishwillee** [comment]
@Elchi3 et al, this is ready for review (when you're back from holiday)

File: files/en-us/web/api/webtransport/webtransport/index.md (hunk 1 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransport/webtransport/index.md:45]: OK, so my assumption was correct - from: https://github.com/w3c/webtransport/issues/508#issuecomment-1539410031

> Typically, there is a third-party that would act as a trusted intermediary between you.
>
> Imagine, for instance, you have a cloud provider that can provision VMs for you. You can ask the provider to launch a virtual machine for you; the cloud provider would create a VM with a WebTransport-based SSH/RDP-like server for its certificate; the server would generate a certificate automatically, and then the cloud provider (having trusted connectivity to the VM) would connect to the server, get its certificate and report its hash via the cloud console. This means that the client now can connect directly to the VM in question via the web UI despite it not having a long-lived TLS certificate (most existing solutions require a Web-to-SSH proxy or something similar).

File: files/en-us/web/api/webtransportreceivestream/getstats/index.md (hunk 9 above)
Comment (hamishwillee) [reply files/en-us/web/api/webtransportreceivestream/getstats/index.md:16]: Verified

File: files/en-us/web/api/webtransport/webtransport/index.md [hunk 13]
```diff
@@ -6,59 +6,111 @@ page-type: web-api-constructor
...
+      - : An array of objects, each defining the hash value of a server certificate along with the name of the algorithm that was used to generate it.
+        This option is only supported for transports using dedicated connections (`allowPooling` is `false`).
+
+        If specified, the browser will attempt to authenticate the certificate provided by the server against the provided certificate hash(es) in order to connect, instead of using the Web public key infrastructure (PKI).
+        If any hashes match, the browser knows that the server has possession of a trusted certificate and will connect as normal.
+        If empty the user agent uses the same PKI certificate verification procedures it would use for a normal fetch operation.
+
+        This feature allows developers to connect to WebTransport servers that would normally find obtaining a publicly trusted certificate challenging, such as hosts that are not publicly routable, or ephemeral hosts like virtual machines.
+
+        > **Note:** The web application might typically fetch the hashes from a trusted intermediary.
```
Comment (hamishwillee) [inline files/en-us/web/api/webtransport/webtransport/index.md:54]: FYI I have tried to update this to capture a few points:
1. The web app never sees a certificate. The browser hashes the downloaded certificate provided during TLS handshake with the indicated hash algorithm. If they match it knows the server has possession of the appropriate certificate and then proceeds to share secrets and set up connection.
2. The app gets the keys via a trusted intermediary. In this case a cloud provider. The cloud provider may need to be trusted via PKI.

**Comment by hamishwillee** [comment]
@Elchi3 This is ready for review. I'm being drip-fed clarifications by the spec authors and developers, but I'm pretty sure that while this might be improved, it is "not wrong". Good to get it in.

**Review by Elchi3 — APPROVED** [review]
Thanks Hamish, looks good to me and I'm happy to merge this. Thanks for double checking with the spec authors, I'm not familiar enough with this API to give it a proper tech preview. Let's get this in. (You need to follow up and add `reliability` to BCD)

**Comment by fippo** [comment]
this needs another look, e.g. it mentions a timestamp on getStats() which was removed in https://github.com/w3c/webtransport/issues/550

@jan-ivar ^

**Comment by hamishwillee** [comment]
> this needs another look, e.g. it mentions a timestamp on getStats() which was removed in [w3c/webtransport#550](https://github.com/w3c/webtransport/issues/550)
> 
> @jan-ivar ^

@fippo Thanks, but please don't necrobump. Can you create specific issues for anything you see as needing fixing - that way any docs work can be shared.

PS I know some it is up to date. I merged some fixes to `WebTransport.draining` last week.
